clang/docs/SYCLFERFCIntegrationHeaderFooter.md

```diff
@@ -0,0 +1,148 @@
+Discourse topic details:
+- Category: [Clang Frontend](https://discourse.llvm.org/c/clang/6)
+- Title: "RFC: SYCL host compiler integration header and footer"
+- Tags: sycl
+
+# Introduction
+SYCL source files are logically (and usually physically) compiled multiple times; code that runs on the "host" (e.g., a CPU) is compiled by a host compiler and code that runs on a "device" (e.g., a GPU) is compiled by a device compiler. The SYCL language is designed such that it is not necessary that the host compiler be SYCL-aware; an ISO C++ conforming compiler suffices. However, some SYCL features such as kernel invocation and support for
+[specialization constants](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_specialization_constants)
+require coordination between the host and device compilers. Integration headers and footers produced by a device compiler are used to communicate the information needed during host compilation to support these features.
+
+# Integration Header
+An integration header is a source file produced during device compilation that is intended to be pre-included during host compilation of the same source file.
+
+Consider the following code:
+
+```
+#include <sycl.hpp>
+
+int main() {
+  sycl::queue q;
+  sycl::buffer<char> b{sycl::range{1024}};
+
+  q.submit([&](sycl::handler &cgh) {
+    sycl::accessor acc{b, cgh};
+    int i = 13;
+    struct S {
+      char c;
+      int i;
+    } test_s;
+    test_s.c = 14;
+
+    cgh.single_task([=] {
+      if (i == 13 && test_s.c == 14) {
+        acc[0] = 'a';
+      }
+    });
+  });
+}
+```
+
+Here the lambda passed to the `kernel_single_task` construct needs to be executed on the device.  The corresponding function object looks like:
```

Comment:
@sergey-semenov, could you please suggest what you would like to see added here?

Reply:
Something along the lines of "This assumes that the layout of the kernel lambda is the same during host and device compilation, which might not be the case when using a different host compiler". Probably makes sense to add it in the section that describes offset & size of lambda fields.

A common example of such a problem is lambdas with captured constexpr variables when using gcc as the host compiler, clang handles them differently.

Reply:
Ah, I now realize where I went wrong with my thoughts on objects of class type with different shape. The host and the device compilers are not aware of the difference in representation for the other. Marshaling can handle differences in size/alignment due to differences in the size/alignment of subobjects, but it can't handle differences due to the presence/absence of subobjects.

Reply:
> A common example of such a problem is lambdas with captured constexpr variables when using gcc as the host compiler, clang handles them differently.

If this is the case, it seems like our third-party host compiler support is probably broken for kernels that capture a constexpr variable.  We should create a document that lists the restrictions when using a third-party host compiler.

Reply:
I think that it's even worse with MSVC.